compose: Deflake testComposeCompare

Xiang-Gu opened the pull request:
We identified the fixed the following issues concerning test testComposeCompare:
1. Disallow several functions because they're not supported in Postgist
2. Disable random generation of `OID` type because it's natural to have the same OIDs assigned to different objects between CRDB and Postgist.
3. Disable using index hints because Postgist does not support this feature.
4. Disable locales because Postgist requires them to be double-quoted
and it's not feasible to change locale name formatting to be double-quoted.
5. Fixed a test bug where we'd mistakenly allow generation of stmts like `CREATE TABLE t (... inverted index (...))` in our test against Postgist.
6. Make this test a "enormous" test which has an 1-hour timeout because it executes two subtests sequentially for 10m each.

Inform #82867
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107224)
<!-- Reviewable:end -->

Xiang-Gu committed:
compose: Deflake testComposeCompare

We identified the fixed the following issues concerning test
testComposeCompare:

1. Disallow several functions because they're not supported in Postgist
2. Disable random generation of OID type because it's natural to have the same OIDs assigned to different objects between CRDB and Postgist.
3. Disable using index hints because Postgist does not support this feature.
4. Disable locales because Postgist requires them to be double-quoted
and it's not feasible to change locale name formatting to be double-quoted.
5. Fixed a test bug where we'd mistakenly allow generation of stmts like
`CREATE TABLE t (... inverted index (...))` in our test against Postgist.
6. Make this test a "enormous" test which has an 1-hour timeout because
it executes two subtests sequentially for 10m each.

Inform #82867
Epic: None
Release note: None

rafiss commented on pkg/sql/randgen/mutator.go:
```diff
@@ -742,10 +742,21 @@ func postgresCreateTableMutator(
+					// Postgres (e.g. `i TEXT COLLATE de_DE` won't work with
+					// Postgres).
+					if colType.Family() == types.CollatedStringFamily {
+						colType.InternalType.Family = types.StringFamily
```
i think this would be simpler if it was:
```
if colType.Family() == types.CollatedStringFamily {
  colType = types.String
}
```

Xiang-Gu replied:
done

rafiss commented on pkg/sql/randgen/mutator.go:
```diff
@@ -742,10 +742,21 @@ func postgresCreateTableMutator(
+					colType := tree.MustBeStaticallyKnownType(def.Type)
+					// Disable locales because CRDB random generator generates
+					// locales without double-quotes, and they are unknown to
+					// Postgres (e.g. `i TEXT COLLATE de_DE` won't work with
```
another alternative is to always add double quotes here, i think:
```
// Format implements the NodeFormatter interface.
func (node *CollateExpr) Format(ctx *FmtCtx) {
	exprFmtWithParen(ctx, node.Expr)
	ctx.WriteString(" COLLATE ")
	lex.EncodeLocaleName(&ctx.Buffer, node.Locale)
}
```

Xiang-Gu replied:
It will break a lot of existing tests, and also need to replace "_" with "-" to make it work. It's too risky and complicated so let's just disable it.

rafiss reviewed (approved):
lgtm, but i had two suggestions

DrewKimball reviewed (commented):
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107224)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @Xiang-Gu)

___
*[`-- commits` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107224#-N_p0XhY8SYLZRYdAmNf):*
Where was this changed?


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107224)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`-- commits` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107224#-N_p0XhY8SYLZRYdAmNf):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Where was this changed?
</blockquote></details>

Look at where the bool "changed" is moved outside the `if !inverted` block for its caller (`ApplyString`) to consume.


<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
TFTR!

bors r+
